### ISSUE TITLE
Proposal to add option excluding audio tracks in specific formats

### ISSUE THREAD

**Issue by JMoVS** [lisamelton/video_transcoding#88]
When faced with a very large collection of DVDs (say 500), I want to automate as much as possible. Right now, I can provide reasonable defaults and get DVDs to probably transcode to a result with good-enough settings. One thing that would be tremedously helpful though is if I could supply a `--only-audio-codec-source AC3`.

This would help in the numerous cases where a DTS and a Dolby AC3 audio track is present on the DVD. This way I could supply an `--add-audio eng,deu` and still make sure that the non-needed audio-codecs in the source won't get picked up.

**Comment by JMoVS** [comment]
another way to go would be an audio option `--no-DTS`

**Comment by JMoVS** [comment]
I tried to achieve my desired goal by using `-H audio-copy-mask=aac,ac3,mp3` but I guess as transcode-video is already passing in the DTS source, there's no way to exlcude it with a handbrake option from the output alltogether...

**Comment by donmelton** [comment]
@JMoVS Are your DVDs ripped to individual `.mkv` files or are they stored as disk image directories?

**Comment by JMoVS** [comment]
They are individual mkv files created with makemkv

**Comment by donmelton** [comment]
@JMoVS OK, so you intentionally left the DTS tracks in there. Probably to be thorough, right? :) I was just curious.

This is a pretty geeky feature request and I really see it only being useful when transcoding DVDs and not Blu-ray Discs. But maybe we can make this work.

Perhaps it's better to design the API to exclude specific formats instead of only including others? Maybe something like this?

```
    --exclude-audio-format FORMAT
```

... where the format string is `dts`, `ac3`, etc. And it could be called multiple times.

**Comment by JMoVS** [comment]
@donmelton Yes, being thorough just in case. ;-)

I have so far no experience with transcoding Blue-Rays, so I'll take your word for it.

Your proposal sounds solid, although I think it's easier to do something like `--exclude-audio-format FORMAT[,FORMAT]` as in one can provide a comma-separated list akin to existing audio options and thereby could potentially mention more than one. Whatever is easier to implement.

**Comment by donmelton** [comment]
@JMoVS Yeah, I can probably make it a list. It's a little more work to parse up front but the implementation would be the same in the core.

**Comment by JMoVS** [comment]
@donmelton I guess as the use-case is very very narrow, if the implementation of calling ot more than once is easier, then go for that. ;-)

**Comment by donmelton** [comment]
@JMoVS This is turning out to be much harder to implement than I thought do the rather narrow-focus of and optimizations within my audio code. And it's likely to take a significant re-architecture to complete. And that has me worried since it will take a lot of testing to make sure I don't break anything.

So, it's not looking like it will go into the release next week. Sorry about that.

**Comment by JMoVS** [comment]
@donmelton M yline of thought here is that both from a workflow and a logical point of view, it could make sense to only call this option as the very last audio option. 

1. Choose and add audio
2. Now remove unwanted codecs

This to mee seems implementation-wise easier. If you store the current audio selection some sort of list/array, later callingf this option could maybe simply remove the eg. DTS audio tracks from said Array before it gets passed into handbrake.

**Comment by donmelton** [comment]
@JMoVS Yes, but what happens if the user places the `--exclude-audio-format` option first?

I'm starting to think that semantically this option does not make sense. For example, my API allows specifying the main track and adding other tracks by their index number. What if one of these very specifically identified tracks is then excluded? Which option wins? By order on the command line? And which track would then become the main track?

BTW, the tracks are already stored in an array. That's not the hard part. No, the hard part is fitting the exclusion into all the other special cases in the `prepare_audio` method. You can't just do it at the end.

**Comment by JMoVS** [comment]
SHAME ON THE USER! 

Kidding aside, I think that this is indeed a problem. Maybe going back to my original suggestion, I thi k this would be more in line with current strategy and also wouldn't give you the "what wins when" thing. A user could supply a `--only-source-audio-with-codec AC3,stereo` and then would achieve the same result but wouldn't have this strange excluding behaviour.

**Comment by JMoVS** [comment]
This would also have the nice benefit that thisnexact pattern could also be used with a "noDTS" option supplied. But it's still a thing on what to **include** and not **exclude**. and that's how I saw the already existing options.

**Comment by donmelton** [comment]
@JMoVS But an option and argument like `--only-source-audio-with-codec AC3,stereo` would raise all the same issues. It's just exclusion by another name.

However, looking at things more closely, the current implementation does allow the last conflicting option to "win" so maybe this is possible. But it's going to take me awhile to settle on the semantics exactly.

**Comment by JMoVS** [comment]
@donmelton Well, that's correct. My mental model is sort of different with the --only-source..., as this suggests to me that this is the very first option that is used when looking at audio-tracks and only the remainings of that will be used. That also seems to help with a lot of conflicts where eventually, the user can reasonably be blamed.

**Comment by donmelton** [comment]
@JMoVS OK, this feature is becoming a pain in the ass. :)

First, regarding the order of options problem. After much study of the existing code (you wouldn't thinks this is hard since I wrote it, but ...:) the only way to implement this feature is to do the exclusion right before actual audio processing and not earlier during argument parsing. In fact, it's not even possible during argument parsing since the input media hasn't even been scanned. This means the exclusion option will override all other audio selection options no matter their order on the command line.

Second, regarding audio format specification. It turns out that HandBrake's nice, clean list of audio output format specifiers, i.e. `aac`, `ac3`, `eac3`, `truehd`, `dts`, `dtshd`, `mp3`, `vorbis`, `flac` and `opus`, don't map very well to the names returned by scanning the input. For example, there's at least three kinds of DTS-HD out there. This is because "libav" is doing most, but not all, of the actual work of parsing the tracks. I'm not even sure yet if the names are consistent across media types.

So, I starting to think this option, if I implement it all, might be reduced in scope to just `--exclude-dts` since that's what you wanted in the first place. :)

Thoughts?

**Comment by JMoVS** [comment]
@donmelton Well, it turns out that I imagined some sort of "before you handle audio at all, pls just exclude DTS".

So I am not sure whether I fully understand it. So let's maybe go through some edge cases:

**Case 1: DVD has Dolby and DTS track in my language**

In this case, if I remember past conversations with you correctly, it doesn't make much sense to include the DTS track. So an option to automatically choose the Dolby track when it comes to the question "What surround track shall I include" would be nice.

**Case 2: DVD only has a DTS surround track in my language (no Dolby)**

In this case, I guess it's better to include the DTS track as the surround  sound. So excluding DTS at the very beginning seems like it's not helping in this very edgy corner case.

**Case 3: Audio setups in Blue-Rays**

I have no personal experience with Blue-Ray Audio formats and what to use and what not to use. So that's for you to imagine the edge cases. ;-)

**Comment by donmelton** [comment]
@JMoVS OK, let's talk about your edge cases:

For your first case, yes, including DTS in output isn't a good idea since playback compatibility can be reduced significantly, especially on Apple devices. (And, yes, I've said this for years. :) However, "automatically" choosing the AC-3 track is not possible because how would `transcoder-video` know which AC-3 track to select? Just because a video contains DTS and AC-3 tracks, even if they're in the same language and channel configuration, doesn't mean they contain the same audio. The only solution is exclusion by format but that turns out to be a blunt instrument, because...

...of your second case. And there's likely many discs in that configuration. Maybe not so "edge" after all, either. So, it seems to me that this option, if I implement it, could only be used selectively. Probably by inspecting each video first. And if a user is going to that trouble up front, they can simply select audio by track index instead and achieve the same result. We don't need a new API to confuse them.

As for Blu-ray Discs, no one has asked me about format exclusion yet. In fact, most of the requests I get are about supporting TrueHD and DTS-HD, meaning those folks want to include the original audio or even transcode to those formats. Which is a really foolish idea but, hey, it's also why I implemented the `--copy-audio` option to keep them quiet for now. :)

Anyway, I'm less inclined to implement this feature. Mostly because now I don't think it's really any more convenient than selecting by track index. It just doesn't seem worth the complication at this point.

**Comment by JMoVS** [comment]
Hmm. Edge case 1 for me always was "same audio track in both formats on this DVD". Never had anything different. So for me, even if this isn't the case, the vast majority of disks are that way. Would improve my defaults and reduce manual handling need a Lot. To be honest, edge case 1 would be the primary use case for me (as audio tracks all were identical in the past for me). 

About the second edge case: It is a theoretical one - I never stumbled upon it. Neither on old DVDs (like from 2000) nor today's DVDs. So I am not really concerned about edge case 2. If I catch 1 disk with edge case 2 out of 300, that's fine. 
As explained, for me they are edge cases to consider. But I have yet to stumble upon a DVD that only contains a DTS track and NOT contain an AC3 track. 

So your proposed 'exclude-DTS' option sounds super awesome for me as it would catch like 99% of my cases perfectly. ;-)

**Comment by lisamelton** [comment]
I've obviously decided not to do this. 🤦‍♀️ I mean, today's simplification and rewrite kinda goes against this design.
